python_modules/dagster/dagster/_core/definitions/op_invocation.py

```diff
@@ -359,13 +361,13 @@ def _key_for_result(result: MaterializeResult, context: "BaseDirectExecutionCont
         return next(iter(context.per_invocation_properties.assets_def.keys))
 
     raise DagsterInvariantViolationError(
-        "MaterializeResult did not include asset_key and it can not be inferred. Specify which"
+        f"{result.__class__.name} did not include asset_key and it can not be inferred. Specify which"
         f" asset_key, options are: {context.per_invocation_properties.assets_def.keys}"
     )
 
 
 def _output_name_for_result_obj(
-    event: MaterializeResult,
+    event: Union[MaterializeResult, ObserveResult],
```

Comment:
This makes me think we should add a marker interface `AssetResult` that both `MaterializeResult` and `ObserveResult` implement for code paths like this

Reply:
One reason I didn't do something like this is because there is also `AssetCheckResult` which would not be included so is kind of confusing, but I'm open to it-- do you think that's preferable to a common base class?